### PR TITLE
Gromacs Node

### PR DESCRIPTION
**Pull request by PythonFZ**
This PR implements a Node for classical sampling of configurations using https://github.com/alanwilter/acpype and the GAFF as well as PACKMOL and RDKit for generating boxes from smiles.

--- 
I've added some trajectory conversion utilities. The energy reading code works in principle but I am fairly certain that the energies I obtain are wrong in some way. There is also an issue that the gmx energy command works on the CLI but not from within the node so I have opted to set the energies to 0 for now.
---

# TODO
- [x] remove `maxwarn`, currently identifiers don't seem to be unique
- [x] box size
- [x] packmol pbc issue
- [ ] make energy / force saving to ASE atoms optional
- [ ] configure gromacs mpi/omp parameters for better performance
- [ ] clean up files? When converting to ASE we have a large duplication w.r.t to the gromacs output files. Do we want to keep them?
- [ ] correct time step in h5md
- [ ] gf2bf (yaml to gromacs) parser

### CODE REVIEW

File: ipsuite/configuration_generation/gmx.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            "0",
+            "|",
+            "gmx",
+            "editconf",
+            "-f",
+            self.box,
+            "-bt",
+            "cubic",
+            "-d",
+            "1",
```
Comment (M-R-Schaefer) [inline ipsuite/configuration_generation/gmx.py:198]: the supercell is already in the self.box file right? is it correct to then use -d 1? does that not lower the density?

File: ipsuite/configuration_generation/gmx.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                mdp_file.resolve().as_posix(),
+                "-c",
+                "box.gro",
+                "-p",
+                "box.top",
+                "-o",
+                "box.tpr",
+                "-v",
+                "-maxwarn",
+                "10",
```
Comment (M-R-Schaefer) [inline ipsuite/configuration_generation/gmx.py:257]: maxwarn 10 seems a bit excessive. Has this been an issue in the IL tests so far?

**Review by M-R-Schaefer — CHANGES_REQUESTED** [review]
some questions on the hardcoded gmx parameters. Otherwise it looks good.

File: ipsuite/configuration_generation/gmx.py (hunk 1 above)
Comment (PythonFZ) [reply ipsuite/configuration_generation/gmx.py]: No, it is not - I'm still struggling with the correct box size

File: ipsuite/configuration_generation/gmx.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        output_file.write_text(atomtypes_section_text)
+
+        # Remove the atomtypes section from the original content
+        modified_content = atomtypes_regex.sub("", content)
+
+        # Write the modified content back to the input file
+        input_file.write_text(modified_content)
+
+
+class Smiles2Gromacs(base.IPSNode):
```
Comment (M-R-Schaefer) [inline ipsuite/configuration_generation/gmx.py:134]: Perhaps a different name would be more informative. To me Smiles2Gromacs sounds like it creates gromacs input files but it also runs simulations.

File: ipsuite/configuration_generation/gmx.py (hunk 1 above)
Comment (M-R-Schaefer) [reply ipsuite/configuration_generation/gmx.py]: hm curious. What is the issue? If packmol can create a box with a correct density, then it should be fine to set -d 0, right? Alternatively, the node can run arbitrary mdp files, we can just add an NPT equilibration step and maybe add a warning about this in the docstring.

File: ipsuite/configuration_generation/gmx.py (hunk 3 above)
Comment (M-R-Schaefer) [reply ipsuite/configuration_generation/gmx.py:154]: Smiles2GromacsRun?

File: ipsuite/configuration_generation/gmx.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                "box.tpr",
+                "-v",
+                "-maxwarn",
+                "10",
+            ]
+            print(f"Running {' '.join(cmd)}")
+            subprocess.run(cmd, check=True, cwd=self.output_dir)
+
+            cmd = ["gmx", "mdrun", "-v", "-deffnm", "box"]
+            subprocess.run(cmd, check=True, cwd=self.output_dir)
```
Comment (M-R-Schaefer) [inline ipsuite/configuration_generation/gmx.py]: is this correctly using the final configuration of one run as the starting configuration of the next?

**Review by M-R-Schaefer — CHANGES_REQUESTED** [review]
minor clarification

File: ipsuite/configuration_generation/gmx.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                "-p",
+                "box.top",
+                "-o",
+                "box.tpr",
+                "-v",
+            ]
+            print(f"Running {' '.join(cmd)}")
+            subprocess.run(cmd, check=True, cwd=self.output_dir)
+            cmd = ["gmx", "mdrun", "-ntmpi", "1", "-v", "-deffnm", "box"]
+            subprocess.run(cmd, check=True, cwd=self.output_dir)
```
Comment (M-R-Schaefer) [inline ipsuite/configuration_generation/gmx.py:280]: There is currently no conversion to ASE atoms right?

Comment (PythonFZ) [reply ipsuite/configuration_generation/gmx.py:280]: https://github.com/zincware/IPSuite/pull/308

**Comment by PythonFZ** [comment]
gf2bf

**Review by M-R-Schaefer — APPROVED** [review]

File: ipsuite/configuration_generation/gmx.py (hunk 3 above)
Comment (PythonFZ) [reply ipsuite/configuration_generation/gmx.py:234]: I think the name is fine - it generates the gromacs input files and once you define the node you need to give the mdp files which clarifies that it is also executing a gromacs run. I don't think we will ever add a Node that only creates the gromacs input files without running them.

**Pull request merged**